test/NuGet.Core.Tests/NuGet.Packaging.Test/ReferencesInNuspecMatchRefAssetsRuleTests.cs

```diff
@@ -0,0 +1,280 @@
+            };
+            var files = new string[]
+            {
+                "ref/net462/MyLib.dll",
+                "ref/net462/MyHelpers.dll"
+            };
+
+            //Act
+            var rule = new ReferencesInNuspecMatchRefAssetsRule();
+            var missingItems = rule.Compare(references, files);
+
+            //Assert
+            Assert.Equal(missingItems.Count(), 1);
+            var singleIssue = missingItems.Single(t => t.TFM.Equals("net462") && t.MissingItems.Contains("MyHelpers.dll"));
+        }
+
+        [Fact]
+        public void Compare_PackageWithRefAssetsUnderRefFolder_ShouldBeEmpty()
+        {
+            //Arrange
+            var references = new Dictionary<string, string[]>();
+            var files = new string[]
+            {
+                "ref/MyLib.dll",
+                "ref/MyHelpers.dll"
+            };
+
+            //Act
+            var rule = new ReferencesInNuspecMatchRefAssetsRule();
+            var missingItems = rule.Compare(references, files);
+
+            //Assert
+            Assert.Empty(missingItems);
+        }
+
+        [Fact]
+        public void Compare_PackageWithAssetsUnderMiscellaneousFolders_ShouldBeEmpty()
+        {
+            //Arrange
+            var references = new Dictionary<string, string[]>();
+            var files = new string[]
+            {
+                "ref/MyCompany/MyLib.dll",
+                "ref/MyCompany/MyHelpers.dll"
+            };
+
+            //Act
+            var rule = new ReferencesInNuspecMatchRefAssetsRule();
+            var missingItems = rule.Compare(references, files);
+
+            //Assert
+            Assert.Empty(missingItems);
+        }
+
+        [Fact]
+        public void Compare_PackageWithMismatchedReferencesInMultipleSubFolders_ShouldHaveCorrectItems()
+        {
+            //Arrange
+            var references = new Dictionary<string, string[]>
+            {
+                {"net462", new string[] {"MyLib.dll", "MyHelpers.dll"} },
+                {"net472", new string[] { "MyLib.dll"} }
+            };
+            var files = new string[]
+            {
+                "ref/net462/MyLib.dll",
+                "ref/net472/MyLib.dll",
+                "ref/net472/MyHelpers.dll"
+            };
+
+            //Act
+            var rule = new ReferencesInNuspecMatchRefAssetsRule();
+            var missingItems = rule.Compare(references, files);
+
+            //Assert
+            Assert.Equal(missingItems.Count(), 2);
+            var singleRefFilesMissing = missingItems.Single(t => t.TFM == "net462" && t.MissingItems.Count() == 1 && t.MissingItems.Contains("MyHelpers.dll"));
+            var singleReferencesMissing = missingItems.Single(t => t.TFM == "net472" && t.MissingItems.Count() == 1 && t.MissingItems.Contains("MyHelpers.dll"));
+        }
+
+        [Fact]
+        public void GenerateWarnings_PackageWithMatchingAssetsAndReferences_ShouldNotWarn()
```

Comment:
if you have time: All the GenerateWarnings_* tests would be easier to understand if you new up a `MissingReference` collection, instead of calling `Compare`.